### PR TITLE
Feat/gefen x optimizer

### PR DESCRIPTION
**Pull request by thad0ctor**
## Description

Adds the [Gefen-X](https://github.com/thad0ctor/Gefen-X) optimizer to axolotl — a memory-efficient AdamW replacement that quantizes optimizer state to ≈1 byte/parameter (≈8× smaller than AdamW's fp32 moments, below bitsandbytes 8-bit) at AdamW-level loss. Builds on the draft [#3755](https://github.com/axolotl-ai-cloud/axolotl/pull/3755) but targets the Gefen-X fork and wires up both variants plus their levers:

- **`optimizer: gefenx`** — core quantized AdamW replacement (`Gefen`).
- **`optimizer: gefenx_muon`** — whole-model Muon hybrid (`GefenMuonHybrid`): Muon updates on 2D hidden weights, Gefen on embeddings / LM head / norms / biases, at the same ≈1 B/param footprint.

All extra constructor levers (`fused`, `factored_v_2d`, `backup_lr`, `sharded_mode`, …) pass through `optim_args`. `gefenx_muon` applies the fork's recommended recipe by default (`backup_lr = 0.5 × lr`, `backup_1d_period_one`, `adjust_lr_fn="match_rms_adamw"`), all overridable.

## Motivation and Context

Full fine-tuning is optimizer-state-bound. Gefen delivers **the lowest peak VRAM and lowest optimizer state in the field at AdamW-level loss and ≈0.96× AdamW throughput** — letting fine tune runs fit on smaller cards without the loss/throughput penalty of 4-bit optimizers. It slots cleanly into axolotl's existing custom-optimizer machinery.

## Basic usage

```yaml
# gefenx — drop-in AdamW replacement (recommended)
optimizer: gefenx
learning_rate: 6.0e-6      # ≈0.6× your AdamW LR (Gefen takes larger effective steps)
bf16: true
optim_args:
  fused: true             # ≈2× faster opt.step, bit-exact (default)
  factored_v_2d: true     # Adafactor-style 2D 2nd moment — matches AdamW loss (default)
```

```yaml
# gefenx_muon — whole-model Muon hybrid, same memory footprint
optimizer: gefenx_muon
learning_rate: 5.0e-5      # Muon tolerates a hotter LR than plain gefenx
bf16: true
optim_args:
  fused: true
  backup_lr: 2.5e-5       # defaults to 0.5 × learning_rate if unset
  sharded_mode: exact     # "distributed" splits Newton-Schulz across GPUs under FSDP2
```

> **Install the `gefenx` extra** — `pip install "axolotl[gefenx]"` (or `pip install gefen-x`), which pulls the [`gefen-x`](https://pypi.org/project/gefen-x/) package; it's also included in the `optimizers` extra bundle. Not `pip install gefen`, which is *upstream* Gefen and lacks this fork's modern-arch memory fallback, fused kernels, and the Muon hybrid.

## Screenshots

**Qwen3-1.7B FFT — loss / speed / memory.** Gefen (fused) sits on the AdamW loss cluster (1.226 vs 1.217) at ≈0.96× throughput, with the lowest peak VRAM (9.21 GiB vs 14.00) and lowest optimizer state (1.01 vs 4.00 B/param).

Qwen3-1.7B optimizer comparison

**Throughput vs peak VRAM frontier** (upper-left = faster + lighter). Gefen (fused) is the frontier point — near-AdamW speed at ⅔ the memory.

Qwen3-1.7B throughput vs VRAM

## How has this been tested?

- **Unit** (`tests/core/test_builders.py`): `gefenx` and `gefenx_muon` build the correct optimizer through the real builder → `create_optimizer()` path; `optim_args` string→type coercion.
- **E2E** (`tests/e2e/test_optimizers.py`): short FFT runs for both variants on `SmolLM2-135M`.
- **Manual**: forward/backward/`step()` on CPU (`fused=False`) confirms both optimizers update parameters and that the hybrid recipe LRs land (muon 1e-3 + backup 5e-4). Verified on RTX 3090 (sm_86), torch 2.12 / cu133, gefen 0.2.0.
- Pinned `ruff` / `ruff-format` / `mypy` / `bandit` clean; generated CLI config options up to date.

## AI Usage Disclaimer

Opus 4.8 used throughout

## Types of changes

- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix
- [ ] Breaking change
- [x] Documentation

## Summary by CodeRabbit

* **New Features**
  * Added support for two new custom optimizer options, including a hybrid variant.
  * Added configuration handling for common optimizer settings and automatic type parsing from text-based values.

* **Documentation**
  * Expanded optimizer docs with setup guidance, recommended settings, and a configuration example.

* **Bug Fixes**
  * Improved how optimizer settings are applied so supported options behave consistently across configurations.

### CODE REVIEW

**Comment by thad0ctor** [comment]
Pushed `3b00f080` (rebased on the latest branch tip): `gefen-x` is now on [PyPI](https://pypi.org/project/gefen-x/). Added an opt-in `gefenx` extra (also included in the `optimizers` bundle) and switched the docs install step to `pip install "axolotl[gefenx]"`. Base install is unchanged (`gefen-x` is CUDA-only and lazily imported). PR description updated to match.

**Comment by thad0ctor** [comment]
resolved  merge errors [81e7105](https://github.com/axolotl-ai-cloud/axolotl/pull/3808/commits/81e7105ad0979cf2c0ed76199a4858a85e53bcdd)
